### PR TITLE
Add mask which displays progress in calibration

### PR DESCRIPTION
**Pull request by DavoliLuca**
Closes: SW-63186

### CODE REVIEW

File: camera_calibration/src/camera_calibration/camera_calibrator.py [hunk 1]
```diff
@@ -229,8 +231,27 @@ def handle_monocular(self, msg):
                                         checkerboard_flags=self.checkerboard_flags,
                                         max_chessboard_speed = self._max_chessboard_speed)
 
-        # This should just call the MonoCalibrator
+        # This should just call the MonoCalibrator and compute the mask for the detected corners if any
+        prec_progress = self._progress
         drawable = self.c.handle_msg(msg)
+        img = self.c.mkgray(msg)
```
Comment (DavoliLuca) [inline camera_calibration/src/camera_calibration/camera_calibrator.py]: @brinij I was thinking on moving all the mask creation into the MonoCalibrator and add the mask as an attribute which could be retrieved, but before that I'd like to get some feedback

**Comment by DavoliLuca** [comment]
Add an overlaying mask red and green on top of the image, it will be green where good corners have been detected, red elsewhere

File: camera_calibration/package.xml [hunk 2]
```diff
@@ -21,7 +21,7 @@
   <run_depend>std_srvs</run_depend>
   <run_depend>sensor_msgs</run_depend>
   
-  <exec_depend>python_orocos_kdl</exec_depend>
+  <run_depend>python_orocos_kdl</run_depend>
```
Comment (brinij) [inline camera_calibration/package.xml:25]: Why did you change it here? exec_depend is new version of run_depend...

File: camera_calibration/src/camera_calibration/camera_calibrator.py (hunk 1 above)
Comment (brinij) [reply camera_calibration/src/camera_calibration/camera_calibrator.py]: Yes please do that, and maybe have an arg of a method the radius of removing the red mask at detected corners.

File: camera_calibration/src/camera_calibration/camera_calibrator.py [hunk 3]
```diff
@@ -229,8 +231,27 @@ def handle_monocular(self, msg):
...
+        scale_height = height / shape[0]
+        scale_width = width / shape[1]
+        if self._red_mask is None:
+            self._red_mask = numpy.uint8(numpy.ones((shape[0],shape[1],1))*(0,0,255))
+        if drawable.params:
+            self._progress = [result[3] for result in drawable.params]
+            if any([old_val < new_val for old_val, new_val in zip(prec_progress, self._progress)]):
+                null_mask = numpy.zeros(shape, numpy.uint8)
+                for corners in self.c.good_corners:
+                    downsampled_corners = [[int(corner[0][0]/scale_height), int(corner[0][1]/scale_width)] for corner in corners[0]]
```
Comment (domeinzer) [inline camera_calibration/src/camera_calibration/camera_calibrator.py]: make tuples directly here: [() for corner in corners[0]]
then you don't need to do it in cv2.circle(null_mask, tuple(el)...

File: camera_calibration/src/camera_calibration/camera_calibrator.py [hunk 4]
```diff
@@ -229,8 +231,27 @@ def handle_monocular(self, msg):
...
+        prec_progress = self._progress
         drawable = self.c.handle_msg(msg)
+        img = self.c.mkgray(msg)
+        height = img.shape[0]
+        width = img.shape[1]
+        shape = drawable.scrib.shape
+        scale_height = height / shape[0]
+        scale_width = width / shape[1]
+        if self._red_mask is None:
+            self._red_mask = numpy.uint8(numpy.ones((shape[0],shape[1],1))*(0,0,255))
```
Comment (domeinzer) [inline camera_calibration/src/camera_calibration/camera_calibrator.py]: let's simplify this a bit: numpy.full((shape[0], shape[1], 3), (0, 0, 255), dtype=np.uint8)

**Review by carzum — COMMENTED** [review]
I like the visualization, reminds me of a mouse nibbling away at some cheese.  But  to me its hard to understand whats going on in the code, there is a lot of code + parameters compressed in a few lines.  I would factor the function out, split it into a few smaller ones and add some argument documentation to them.

File: camera_calibration/src/camera_calibration/camera_calibrator.py (hunk 1 above)
Comment (DavoliLuca) [reply camera_calibration/src/camera_calibration/camera_calibrator.py]: addressed

File: camera_calibration/package.xml (hunk 2 above)
Comment (DavoliLuca) [reply camera_calibration/package.xml:25]: addressed

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 5]
```diff
@@ -846,7 +847,7 @@ def undistort_points(self, src):
             return cv2.undistortPoints(src, self.intrinsics, self.distortion, R = self.R, P = self.P)
         elif self.camera_model == CAMERA_MODEL.FISHEYE:
             return cv2.fisheye.undistortPoints(src, self.intrinsics, self.distortion, R = self.R, P = self.P)
-
+
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py]: remove trailing space

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 6]
```diff
@@ -938,6 +939,22 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+    def update_mask(self, img, screen):
+        """
+        Update or if None create a red mask with the same size of screen
+        and add on top of it detection areas displayed through green circles
+        """
+        height, width = img.shape[0], img.shape[1]
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        if self.mask is None:
+            self.mask = numpy.uint8(numpy.ones((shape[0], shape[1], 1)) * (0, 0, 255))
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py]: Add a comment here that red mask is created here

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 7]
```diff
@@ -938,6 +939,22 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        else:
+            for corners in self.good_corners:
+                downsampled_corners = [
+                    [int(corner[0][0] / scale_height), int(corner[0][1] / scale_width)] for corner in corners[0]]
+                for index, el in enumerate(downsampled_corners):
+                    self.mask = cv2.circle(self.mask, tuple(el), 35, (0, 255, 0), -1)
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py]: please make the radius of 35 as arg for update_mask with the default value of 35

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 8]
```diff
@@ -938,6 +939,22 @@ def pt2line(x0, y0, x1, y1, x2, y2):
         else:
             return None
 
+    def update_mask(self, img, screen):
+        """
+        Update or if None create a red mask with the same size of screen
+        and add on top of it detection areas displayed through green circles
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py]: Add a comment what is this mask used for -> visualization of covered image area with good corner detection
red-> not covered, green->covered
Also you are missing docstring of arguments, please add

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 9]
```diff
@@ -938,6 +939,22 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        height, width = img.shape[0], img.shape[1]
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        if self.mask is None:
+            self.mask = numpy.uint8(numpy.ones((shape[0], shape[1], 1)) * (0, 0, 255))
+        else:
+            for corners in self.good_corners:
+                downsampled_corners = [
+                    [int(corner[0][0] / scale_height), int(corner[0][1] / scale_width)] for corner in corners[0]]
+                for index, el in enumerate(downsampled_corners):
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py]: index is not used, call it _ , and el is too short variable name, please call it element

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 10]
```diff
@@ -938,6 +939,22 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        """
+        height, width = img.shape[0], img.shape[1]
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        if self.mask is None:
+            self.mask = numpy.uint8(numpy.ones((shape[0], shape[1], 1)) * (0, 0, 255))
+        else:
+            for corners in self.good_corners:
+                downsampled_corners = [
+                    [int(corner[0][0] / scale_height), int(corner[0][1] / scale_width)] for corner in corners[0]]
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py]: this line feels too long

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 11]
```diff
@@ -938,6 +939,35 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        :param img: source image
+        :type img: :class:`cvMat`
+
+        :param screen: display
+        :type screen: :class:`cvtColor`
+
+        :param radius: radius of the green circle which shows where corner detection has occurred
+        :type radius: :class:`int`
+
+        Update or if None create a red mask which shows where no corner detection has occurred
```
Comment (carzum) [inline camera_calibration/src/camera_calibration/calibrator.py]: in the reST  doc string format the function description is before the arguments. Return is missing, but its None anyways

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 12]
```diff
@@ -938,6 +939,35 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        green circles which represents instead areas of the image where corners have been detected
+        """
+        height, width = img.shape[0], img.shape[1]
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        # Create mask if none
+        if self.mask is None:
+            self.mask = numpy.uint8(numpy.ones((shape[0], shape[1], 1)) * (0, 0, 255))
+        # Else update it 
+        else:
```
Comment (carzum) [inline camera_calibration/src/camera_calibration/calibrator.py]: why only do it in the else case? I think you always want to update the mask when this is called. This will return an empty mask for the first detection

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 13]
```diff
@@ -938,6 +939,35 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+            for corners in self.good_corners:
+                downsampled_corners = [
+                    [int(corner[0][0] / scale_height), int(corner[0][1] / scale_width)] \
+                     for corner in corners[0]]
+                for _, element in enumerate(downsampled_corners):
+                    self.mask = cv2.circle(self.mask, tuple(element), radius, (0, 255, 0), -1)
```
Comment (carzum) [inline camera_calibration/src/camera_calibration/calibrator.py]: this is technically not a mask, but an RGB image. I would find it cleaner to just use a binary mask in the attribute and then only make it a color image when adding the layers in camera_calibrator.py

**Review by carzum — CHANGES_REQUESTED** [review]

File: camera_calibration/package.xml [hunk 14]
```diff
@@ -14,12 +14,12 @@
...
-  <run_depend>message_filters</run_depend>
-  <run_depend>rospy</run_depend>
-  <run_depend>std_srvs</run_depend>
-  <run_depend>sensor_msgs</run_depend>
+  <exec_depend>cv_bridge</exec_depend>
+  <exec_depend>image_geometry</exec_depend>
+  <exec_depend>message_filters</exec_depend>
+  <exec_depend>rospy</exec_depend>
+  <exec_depend>std_srvs</exec_depend>
+  <exec_depend>sensor_msgs</exec_depend>
```
Comment (domeinzer) [inline camera_calibration/package.xml:21]: would be nice to keep the list sorted: sensor < std

File: camera_calibration/src/camera_calibration/calibrator.py (hunk 6 above)
Comment (domeinzer) [reply camera_calibration/src/camera_calibration/calibrator.py]: let's simplify this a bit: numpy.full((shape[0], shape[1], 3), (0, 0, 255), dtype=np.uint8)

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 15]
```diff
@@ -938,6 +939,35 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        # Create mask if none
+        if self.mask is None:
+            self.mask = numpy.uint8(numpy.ones((shape[0], shape[1], 1)) * (0, 0, 255))
+        # Else update it 
+        else:
+            for corners in self.good_corners:
+                downsampled_corners = [
+                    [int(corner[0][0] / scale_height), int(corner[0][1] / scale_width)] \
```
Comment (domeinzer) [inline camera_calibration/src/camera_calibration/calibrator.py]: make tuples directly here: [() for corner in corners[0]]
then you don't need to do it in cv2.circle(null_mask, tuple(el)...

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 16]
```diff
@@ -938,6 +939,33 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        # Create the binary mask
+        self.mask = numpy.full((shape[0], shape[1], 1), 255, dtype=numpy.uint8)
+        # Update the mask with the corners detected
+        for corners in self.good_corners:
+            downsampled_corners = [
+                (int(corner[0][0] / scale_height), int(corner[0][1] / scale_width)) \
+                for corner in corners[0]]
+            for _, element in enumerate(downsampled_corners):
```
Comment (domeinzer) [inline camera_calibration/src/camera_calibration/calibrator.py]: we should not use enumerate if we ignore the "i" later

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 17]
```diff
@@ -938,6 +939,33 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        :type screen: :class:`cvMat`
+
+        :param radius: radius of the circle which shows where corner detection has occurred
+        :type radius: :class:`int`
+        """
+        height, width = img.shape[0], img.shape[1]
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        # Create the binary mask
+        self.mask = numpy.full((shape[0], shape[1], 1), 255, dtype=numpy.uint8)
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py]: Since detected corners never turn back to not detected, there is no need to create the mask every time you want to update it. Just create it here only when it is None... Please correct me if I am overlooking something...

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 18]
```diff
@@ -937,9 +938,44 @@ def pt2line(x0, y0, x1, y1, x2, y2):
             return math.sqrt(sum([e**2 for e in errors]) / len(errors))
         else:
             return None
+    def create_mask(self, screen):
```
Comment (brinij) [inline camera_calibration/src/camera_calibration/calibrator.py:942]: missing empty line

**Review by brinij — APPROVED** [review]

File: camera_calibration/nodes/cameracalibrator.py [hunk 19]
```diff
@@ -225,7 +228,7 @@ def main():
     rospy.init_node('cameracalibrator')
     node = OpenCVCalibrationNode(boards, options.service_check, sync, calib_flags, fisheye_calib_flags, pattern, options.camera_name,
                                  checkerboard_flags=checkerboard_flags, max_chessboard_speed=options.max_chessboard_speed,
-                                 queue_size=options.queue_size, no_gui=options.no_gui, no_mono_if_stereo=options.no_mono_if_stereo)
+                                 queue_size=options.queue_size, no_gui=options.no_gui, no_mono_if_stereo=options.no_mono_if_stereo, mask = options.mask)
```
Comment (carzum) [inline camera_calibration/nodes/cameracalibrator.py]: the line is 153 chars long, a line break would be more sensible

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 20]
```diff
@@ -938,8 +939,44 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        """
+        Create a full binary mask with the image shape
 
-    def handle_msg(self, msg):
+        :param screen: display
+        :type screen: :class:`cvMat`
+        """
+        shape = screen.shape
+        # Create the binary mask
```
Comment (carzum) [inline camera_calibration/src/camera_calibration/calibrator.py:950]: thats not a binary mask. Why not make it bool?

File: camera_calibration/src/camera_calibration/camera_calibrator.py [hunk 21]
```diff
@@ -229,8 +230,15 @@ def handle_monocular(self, msg):
                                         checkerboard_flags=self.checkerboard_flags,
                                         max_chessboard_speed = self._max_chessboard_speed)
 
-        # This should just call the MonoCalibrator
-        drawable = self.c.handle_msg(msg)
+        # This should just call the MonoCalibrator and compute the mask for the detected corners if any
+        drawable = self.c.handle_msg(msg, self._mask)
+        if self.c.mask is not None and self._mask:
+            # Transforming the mask in a BGR image
+            blue_layer = numpy.zeros(self.c.mask.shape, dtype=numpy.uint8)
```
Comment (carzum) [inline camera_calibration/src/camera_calibration/camera_calibrator.py]: instead of creating 4 different arrays you can just initialize the bgr image with zeros and update the red and green layer with mask * 255 and ~mask*255

File: camera_calibration/src/camera_calibration/camera_calibrator.py [hunk 22]
```diff
@@ -229,8 +230,15 @@ def handle_monocular(self, msg):
                                         checkerboard_flags=self.checkerboard_flags,
                                         max_chessboard_speed = self._max_chessboard_speed)
 
-        # This should just call the MonoCalibrator
-        drawable = self.c.handle_msg(msg)
+        # This should just call the MonoCalibrator and compute the mask for the detected corners if any
+        drawable = self.c.handle_msg(msg, self._mask)
+        if self.c.mask is not None and self._mask:
```
Comment (carzum) [inline camera_calibration/src/camera_calibration/camera_calibrator.py]: i would put the if + everything inside into a separate function

File: camera_calibration/nodes/cameracalibrator.py (hunk 19 above)
Comment (domeinzer) [reply camera_calibration/nodes/cameracalibrator.py]: also no whitespaces around = here

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 23]
```diff
@@ -938,8 +939,44 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        :param radius: radius of the circle which shows where corner detection has occurred
+        :type radius: :class:`int`
+        """
+        height, width = img.shape[0], img.shape[1]
+        shape = screen.shape
+        scale_height, scale_width = height / shape[0], width / shape[1]
+        # Update the mask with the corners detected
+        for corners in self.good_corners:
+            downsampled_corners = [
+                (int(corner[0][0] / scale_height), int(corner[0][1] / scale_width)) \
```
Comment (domeinzer) [inline camera_calibration/src/camera_calibration/calibrator.py]: i don't think you need the  \ at the end of the line here and it should be removed if unnecessary

File: camera_calibration/src/camera_calibration/camera_calibrator.py [hunk 24]
```diff
@@ -138,7 +138,7 @@ def run(self):
 class CalibrationNode:
     def __init__(self, boards, service_check = True, synchronizer = message_filters.TimeSynchronizer, flags = 0,
                 fisheye_flags = 0, pattern=Patterns.Chessboard, camera_name='', checkerboard_flags = 0,
-                max_chessboard_speed = -1, queue_size = 1, no_gui=False, no_mono_if_stereo=False):
+                max_chessboard_speed = -1, queue_size = 1, no_gui=False, no_mono_if_stereo=False, mask = False):
```
Comment (domeinzer) [inline camera_calibration/src/camera_calibration/camera_calibrator.py]: no whitespaces around = here
whitespaces are added when you are using typing, but without, you don't add them.
optionally, you can fix the old values, but at least you should not continue the pattern :)

File: camera_calibration/nodes/cameracalibrator.py [hunk 25]
```diff
@@ -71,6 +71,9 @@ def main():
                      help="calibration pattern to detect - 'chessboard', 'circles', 'acircles', 'charuco'\n" +
                      "  if 'charuco' is used, a --charuco_marker_size and --aruco_dict argument must be supplied\n" +
                      "  with each --size and --square argument")
+    group.add_option("--mask",
+                     action="store_true", dest="mask", default=False,
+                     help="Whether displaying or no an overlaying mask displaying corner detection on top of the image")
```
Comment (carzum) [inline camera_calibration/nodes/cameracalibrator.py]: Grammar is weird here, 'Whether displaying of no'
I would just change it to 
'Set to display a mask with detected corners on top of the image'

File: camera_calibration/src/camera_calibration/calibrator.py [hunk 26]
```diff
@@ -938,8 +939,45 @@ def pt2line(x0, y0, x1, y1, x2, y2):
...
+        """
+        Create a full binary mask with the image shape
 
-    def handle_msg(self, msg):
+        :param screen: display
+        :type screen: :class:`cvMat`
+        """
+        shape = screen.shape
+        # Create the binary mask
+        self.mask = numpy.full((shape[0], shape[1]), 1, dtype=numpy.uint8)
```
Comment (carzum) [inline camera_calibration/src/camera_calibration/calibrator.py]: I dont see a reason to  use uint8 as a dtype.  Its used as a binary mask  and even converted to bool. just use bool

**Review by carzum — CHANGES_REQUESTED** [review]

**Review by carzum — APPROVED** [review]
